To-do 리스트 Redux를 사용하여 리팩터링 하기

2scent opened the pull request:
안녕하세요. 이번 한 주 동안 잘 부탁드립니다.

아샬 님의 강의라는 모범 답안(?)을 보고, 과제를 진행하니 결과물이 비슷하게 나오는 것 같네요.

우선 리덕스 관련 부분부터 TDD로 만들어 봤는데 이런 식으로 진행하면 될까요?

2scent committed:
Add Redux with default settings

redux, react-redux, @reduxjs/toolkit 설치
빈 store 추가하고, index.js에서 Provider로 적용

2scent committed:
Add updateTaskTitle action test

2scent committed:
Implement updateTaskTitle action

2scent committed:
Add addTask action tests

- tasks에 newId를 가진 새로운 task가 추가되는지 테스트
- taskTitle이 빈 문자열이 되는지 테스트
- newId이 newId + 1로 업데이트 되는지 테스트

2scent committed:
Implement addTask action

2scent committed:
Refactor addTask action

2scent committed:
Add addTask action test without task title

2scent committed:
Implement addTask action without task title

2scent committed:
Add deleteTask action test

2scent committed:
Implement deleteTask action

2scent committed:
Add deleteTask test with non-existent task id

2scent committed:
Refactor reducer test

action 정의는 actions.js에 action creator로 분리
중복 코드 제거

2scent committed:
Add reducer tests with unhandled actions

2scent committed:
Fix Uncaught TypeError

상태에 프로퍼티가 없어서 destructure 관련 에러가 발생한 것을 초기 상태를
추가해서 해결

2scent committed:
Add reducer test without state

daadaadaah commented on src/reducer.js:
```diff
@@ -9,6 +9,8 @@ const reducer = (state, action) => {
   if (action.type === 'addTask') {
     const { newId, tasks, taskTitle } = state;
 
+    if (!taskTitle) return state;
```
`if`문에 `{}` 붙여주세요!

daadaadaah commented on src/reducer.js:
```diff
@@ -0,0 +1,41 @@
+const reducer = (state = initialState, action) => {
+  const { newId, taskTitle, tasks } = state;
+
+  if (action.type === 'updateTaskTitle') {
+    return {
+      ...state,
+      taskTitle: action.payload.taskTitle,
+    };
+  }
+
+  if (action.type === 'addTask') {
+    if (!taskTitle) return state;
+
+    return {
+      ...state,
+      newId: newId + 1,
+      taskTitle: '',
+      tasks: [
+        ...tasks,
+        { id: newId, title: taskTitle },
+      ],
+    };
+  }
+
+  if (action.type === 'deleteTask') {
+    return {
+      ...state,
+      tasks: tasks.filter(({ id }) => id !== action.payload.id),
+    };
+  }
```
reducer 안에 `if`이 점점 많아지는 것 같아요!
점점 많아지게 되면, 읽기 어려울 것 같아요!
`if`나 `switch`를 사용하지 않고, 어떻게 이문제를 해결할 수 있을까요?

daadaadaah commented on src/reducer.test.js:
```diff
@@ -0,0 +1,123 @@
+    it("doesn't work", () => {
+      const state = reducer(
+        { taskTitle: '' },
+        {},
```
`{ type: undefined }` 이렇게 명시적으로 표현해주시는건 어떠신가요? 저는 의도를 최대한 드러나게 코드를 작성하는 편이예요

daadaadaah commented on src/reducer.test.js:
```diff
@@ -0,0 +1,123 @@
+
+    context('with non-existent task id', () => {
+      it("doesn't work", () => {
+        const state = reduceDeleteTask(32134);
```
저라면, `-232` 이런 숫자를 non-exixtent task id로 사용할 것 같아요!

daadaadaah reviewed (commented):
너무 잘하셨어요!👍👍👍 TDD 해보니까, 어떠신가요? 감이 좀 오시나요? 
결과물이 아샬님과 비슷한 느낌이 나는건 잘하고 있다는 의미입니다!ㅎㅎ
좋은 패턴이 그대로 익혀지고 있다는 신호거든용!
다음 PR도 기대하고 있을게요😎

2scent committed:
Refactor reducer test

with unhandled actions에서 처리하지 않는 액션이란 것을 명확히 표현
with non-existent task id에서 존재하지 않는 id란 것을 명확히 표현

2scent committed:
Refactor reducer

각 action을 처리하던 부분을 action-handler.js 파일로 분리

2scent committed:
Refactor App, App test to use redux

2scent committed:
Add ListContainer with test

ListContainer와 그 테스트를 추가
App과 Page 컴포넌트에도 ListContainer 적용

2scent committed:
Add InputContainer with test

2scent committed:
Refactor App to render containers directly

App 컴포넌트에서 InputContainer, ListContainer를 직접 렌더하도록 수정
필요 없어진 Page 컴포넌트 삭제

2scent commented:
피드백 주신 부분 참고해서 리덕스 관련 코드 리팩토링 했습니다.

`ListContainer`, `InputContainer` 컴포넌트들 추가해서 `App` 컴포넌트에서 직접 렌더링하도록 하고, 필요 없어진 `Page` 컴포넌트는 삭제했습니다.

`App` 컴포넌트의 테스트는 작성 못하는 등 마지막에 급하게 하느라 부족한 부분이 많을 것 같은데 이번에도 많은 피드백 부탁드립니다 😅

daadaadaah commented on src/action-handler.js:
```diff
@@ -0,0 +1,33 @@
+
+const actionHandler = (action) => (actionHandlers[action.type] || defaultHandler)(action);
+
+export default actionHandler;
```
`action-handler.js` 따로 빼신 이유가 있을까용? `reducer.js` 파일 안에서 관리하는게 유지보수하기 좋을 것 같아서요!

2scent replied:
관련 코드를 `reducer.js` 안에 작성하니깐 좀 복잡해 보여서 나눴는데 지금 다시 보니깐 말씀하신 것처럼 같은 파일에 있는 게 유지보수할 때 좋을 것 같네요.

daadaadaah commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+  });
+
+  it('renders the task title', () => {
+    const { getByRole } = render(<InputContainer />);
```
중복되는 코드 리팩터링 한번 해보시는거 어때요?
테스트코드도 다 작성하신 후 리팩터링할거 없나 한번 생각해보시는게 좋아요!

daadaadaah commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+  it('renders the task title', () => {
+    const { getByRole } = render(<InputContainer />);
+
+    expect(getByRole('textbox')).toHaveDisplayValue(taskTitle);
```
`getByRole`보다 좀더 구체적으료 표현할 수 있는 것은 없을까요?
이것만 봤을 때, 어떤 textbox인지 의문이 생길 것 같아요!

daadaadaah commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+  it('dispatches a addTask action when the button is clicked', () => {
+    const { getByRole } = render(<InputContainer />);
+
+    fireEvent.click(getByRole('button'));
```
`getByRole`보다 좀더 구체적으료 표현할 수 있는 것은 없을까요?
이것만 봤을 때, 어떤 button인지 의문이 생길 것 같아요!

daadaadaah commented on src/ListContainer.test.jsx:
```diff
@@ -0,0 +1,45 @@
+  const tasks = [
+    { id: 1, title: '할 일 1' },
+    { id: 2, title: '할 일 2' },
+  ];
```
여러 곳에서 사용되는 tasks 데스트용 데이터를 보통 fixture라고 이야기합니다. 
저는 이러한 데이터를 fixtures 폴더 하나 만들어서 따로 관리하는 편이예요!

daadaadaah commented on src/reducer.js:
```diff
@@ -0,0 +1,11 @@
+  tasks: [],
+};
+
+const reducer = (state = initialState, action) => actionHandler(action)(state);
```
`actionHandlers`를 커링으로 하신 이유가 있으신가요? 좀더 간단한 방법도 있을 것 같아서요!

2scent replied:
`action-handler.js`로 파일을 따로 나눌 때, `reducer.js`에서 `action`과 `state`를 한 번만 넘겨주기 위해서 이렇게 구현했는데 코드를 다시 `reducer.js`로 통합하고, 보다 간단하게 해보겠습니다.

daadaadaah reviewed (commented):
1. 화살표 함수 사용하실 때, `{}`를 붙여주는게 읽기 더 편할것 같아요!
2. 잘 하고 계십니다👍 App 테스트 코드 기대하고 있을게요!

2scent committed:
Refactor reducer

action-handler 삭제
action 처리 관련 코드를 reducer.js에서 관리

2scent committed:
Refactor InputContainer test

- InputContainer 렌더하는 코드 중복 제거
- 쿼리 관련 코드를 보다 명확하게 수정

2scent committed:
Refactor ListContainer test

ListContainer 렌더 코드 중복 제거
쿼리 코드 보다 명확하게 수정

2scent committed:
Add App test

2scent committed:
Add fixtures

fixtures 추가하고, 테스트에 적용

2scent commented:
피드백 주신 것들 반영했고, `App` 컴포넌트 테스트랑 fixtures도 추가했습니다.

`App` 컴포넌트 테스트는 `To-do`라는 제목의 존재를 확인한 테스트 말고는 사실상 `InputContainer` 컴포넌트와 `ListCotainer` 컴포넌트의 테스트를 그냥 복붙한 수준인데 "이런 테스트가 의미가 있을까?"라는 생각이 들었습니다.

fixtures를 추가하고, 필요하다고 느낀 곳에 적용해 봤는데 이렇게 하는 게 맞을까요?

daadaadaah reviewed (approved):
잘 하셨어요!👏 다음 과제 진행해주세요🚀
